telemetry/telemetry.go

```diff
@@ -211,121 +229,207 @@ func isFlagPassed(fs *flag.FlagSet, name string) bool {
 	return found
 }
 
-func signalHandler(serverControlSignal chan<- int, wg *sync.WaitGroup, sigchannel <-chan os.Signal) {
-	defer wg.Done()
+func iNotifyCertMonitoring(watcher *fsnotify.Watcher, telemetryCfg *TelemetryConfig, serverControlSignal chan<- ServerControlValue, testReadySignal chan<- int) {
+	defer watcher.Close()
+
+	done := make(chan bool)
+
+	go func() {
+		if testReadySignal != nil { // for testing only
+			testReadySignal <- 0
+		}
+		for {
+			select {
+			case event := <-watcher.Events:
+				if event.Name != "" && (filepath.Ext(event.Name) == ".cert" || filepath.Ext(event.Name) == ".crt" ||
+                                   filepath.Ext(event.Name) == ".pem" || filepath.Ext(event.Name) == ".key") {
+					log.V(1).Infof("Inotify watcher has received event: %v", event)
+					if event.Op&fsnotify.Write == fsnotify.Write || event.Op&fsnotify.Create == fsnotify.Create {
+						log.V(1).Infof("Cert File has been modified: %s", event.Name)
+						serverControlSignal <- ServerStart // let server know that a write/create event occurred
+						done <- true
+						return
+					}
+					if event.Op&fsnotify.Remove == fsnotify.Remove || event.Op&fsnotify.Rename == fsnotify.Rename {
+						log.V(1).Infof("Cert file has been deleted: %s", event.Name)
+						serverControlSignal <- ServerRestart // let server know that a remove/rename event occurred
+						done <- true
+						return
+					}
+				}
+			case err := <-watcher.Errors:
+				if err != nil {
+					log.Errorf("Received error event when watching cert: %v", err)
+					serverControlSignal <- ServerStop
+					done <- true
+					return // If watcher is unable to access cert file stop monitoring
+				}
+			}
+		}
+	}()
+
+	telemetryCertDirectory := filepath.Dir(*telemetryCfg.ServerCert)
+
+	log.V(1).Infof("Begin cert monitoring on %s", telemetryCertDirectory)
+
+	err := watcher.Add(telemetryCertDirectory) // Adding watcher to cert directory
```

Comment:
I think we don't need monitor the whole directory, and we only care about CA cert and chained cert.

Reply:
fsnotify recommends to not use watcher on specific files but on directories. I think it should be fine as we filter out the file events based off extension that we care about like .pem, .key, .cert, etc.

https://pkg.go.dev/github.com/fsnotify/fsnotify#section-readme